h2o-core/src/main/java/water/rapids/ast/prims/advmath/SpearmanCorrelation.java

```diff
@@ -1,72 +1,132 @@
 package water.rapids.ast.prims.advmath;
 
+import water.Key;
 import water.MRTask;
 import water.Scope;
 import water.fvec.Chunk;
 import water.fvec.Frame;
 import water.fvec.Vec;
-import water.rapids.Env;
 import water.rapids.Merge;
-import water.rapids.Val;
-import water.rapids.ast.AstPrimitive;
-import water.rapids.ast.AstRoot;
-import water.rapids.vals.ValNum;
 import water.util.FrameUtils;
-import water.util.VecUtils;
 
 import java.math.BigDecimal;
 import java.math.MathContext;
+import java.util.Objects;
+
+public class SpearmanCorrelation {
+
+  public static Frame calculate(final Frame frameX, Frame frameY, final AstCorrelation.Mode mode) {
+    Objects.requireNonNull(frameX);
+    Objects.requireNonNull(frameY);
+
+    final Frame correlationMatrix = createCorrelationMatrix(frameX);
+    final boolean framesAreEqual = frameX == frameY; // TODO: Client APIs create new frames with no keys to compare
```

Comment:
how is this todo related to this functionality?

Reply:
Thank you. Fixed.

## Why the TODO ?
This piece of code was effectively not working. The  variable `framesAreEqual` is used for optimization on the diagonal of the correlation matrix - if the vectors compared are guaranteed to be the same, the correlation is automatically `1` without any calculations required. With one exception, which is `AstCorrelation.Mode.Everything` - this mode still requires to inspect all the observations and output NaN if there is an observation with a single NaN found.

Rapids will create two "distinct" frames (object address is distinct), but with the same content in case the correlation is created within one frame. Frame X and Frame Y will contain the same vecs, but object-wise, they'll be different frames. Therefore, `equals` or direct comparison with `==` won't work.

## Resolution

This has been resolved by introducing `framesContainSameVecs`, which compares the vectors in a Frame key-wise. And if they're the same, this optimization can happen.